### PR TITLE
New test configs

### LINKED ISSUE #3258: Net-new test configs
### Describe the feature

Let's add configs for:
- `where`
- `limit`
- `warn_if`, `error_if` (picking up from https://github.com/fishtown-analytics/dbt/issues/2219 + https://github.com/fishtown-analytics/dbt/issues/2982)

Following the work in #3257, these will be configurable from `config()` + `dbt_project.yml` + `.yml` instantiation (for generic tests).

### Mechanism

1. Add to TestConfig:
https://github.com/fishtown-analytics/dbt/blob/7203123d56e42007d569a03c76bea0f351fb1611/core/dbt/contracts/graph/model_config.py#L437-L440

2. Add to schema test builder (along with methods lower down):
https://github.com/fishtown-analytics/dbt/blob/7203123d56e42007d569a03c76bea0f351fb1611/core/dbt/parser/schema_test_builders.py#L187-L188

3. Add to shortcut method for "fast" schema tests:
https://github.com/fishtown-analytics/dbt/blob/7203123d56e42007d569a03c76bea0f351fb1611/core/dbt/parser/schemas.py#L490-L491

### PR DESCRIPTION
**Pull request by kwigley**
reopened from #3336! (thanks @jtcohen6 )

resolves #3258, #3321

### Description

```yml
tests:
  +warn_if: ">10"
  +error_if: ">100"
  +where: "date_col = current_date"
  +limit: 10
  +fail_calc: "count(*)"
```

Adds ~four~ five new test configs:
- `limit`: Simple enough, templated out in the materialization. This is mostly a complement for #3316
- `where`: The way I did this was _unbelievably_ hacky—see the test cases for yourself—but _it works_, and in a way that should be backwards compatible with all existing schema/generic tests, without them needing to change any part of their SQL definition. (This config doesn't make sense for one-off tests.)
- `warn_if`, `error_if`: The user supplies a python-evaluable string (e.g. `>=3`, `<5`, `==0`, `!=0`) and dbt will compare the fail count/calc against it.
    - By default, set to `!=0`
    - Interaction with `severity`: A little tricky, but I actually think they work reasonably together. By default, `severity: error`, and dbt checks the `error_if` condition first; if not error, then check the `warn_if` condition; if the result meets neither, it passes the test. If the user sets `severity: warn`, dbt will skip over the `error_if` condition entirely and jump straight to `warn_if`.
- `fail_calc`: https://github.com/fishtown-analytics/dbt/issues/3321

### Possible follow up work
- Cleanup: Could we wrap the `where` logic in a special internal macro, so long as it's included in the schema test parsing context?

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
   - Need to add some integration tests, non-blocking for review
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by gshank** [comment]
Git is terminally confused by this branch. Somehow a commit of mine (b952b701702f) seems to have gotten into this. I tried rebasing on develop because I like to look at a squashed commit to see all the changes and git couldn't do it. Could you make a clean branch and force push it?

**Comment by kwigley** [comment]
@gshank good call! Looks like I totally messed up a rebase at some point, should be in good shape now!

File: core/dbt/parser/schemas.py [hunk 1]
```diff
@@ -488,11 +488,28 @@ def render_test_update(self, node, config, builder):
...
+                node.config['limit'] = builder.limit()
+            if builder.warn_if() is not None:
+                node.unrendered_config['warn_if'] = builder.warn_if()
+                node.config['warn_if'] = builder.warn_if()
+            if builder.error_if() is not None:
+                node.unrendered_config['error_if'] = builder.error_if()
+                node.config['error_if'] = builder.error_if()
+            if builder.fail_calc() is not None:
+                node.unrendered_config['fail_calc'] = builder.fail_calc()
+                node.config['fail_calc'] = builder.fail_calc()
```
Comment (gshank) [inline core/dbt/parser/schemas.py:512]: I'd prefer to see this in the builder as a method that just returns the configs. But we can leave that for later.

**Review by gshank — APPROVED** [review]
LGTM!

File: test/integration/008_schema_tests_test/test_schema_v2_tests.py [hunk 2]
```diff
@@ -125,6 +124,42 @@ def test_postgres_malformed_schema_strict_will_break_run(self):
...
+    def project_config(self):
+        return {
+            'config-version': 2,
+            "macro-paths": ["macros-v2/custom-configs"],
+        }
+
+    @use_profile('postgres')
+    def test_postgres_config(self):
+        """ Test that tests use configs properly. All tests for
+        this project will fail, configs are set to make test pass. """
```
Comment (jtcohen6) [inline test/integration/008_schema_tests_test/test_schema_v2_tests.py:150]: I like it!

**Review by jtcohen6 — APPROVED** [review]
lgtm!

Happy to merge #3316 before or after, in either case there will be some conflicts.

File: core/dbt/contracts/graph/model_config.py [hunk 3]
```diff
@@ -438,6 +438,34 @@ class SeedConfig(NodeConfig):
 class TestConfig(NodeConfig):
     materialized: str = 'test'
     severity: Severity = Severity('ERROR')
+    where: Optional[str] = None
+    limit: Optional[int] = None
+    fail_calc: str = "count(*)"
```
Comment (iknox-fa) [inline core/dbt/contracts/graph/model_config.py:443]: Minor detail, but s/"/'/

File: core/dbt/parser/schema_test_builders.py [hunk 4]
```diff
@@ -278,6 +280,21 @@ def severity(self) -> Optional[str]:
         else:
             return None
 
+    def where(self) -> Optional[str]:
```
Comment (iknox-fa) [inline core/dbt/parser/schema_test_builders.py:283]: Minor detail:  This is a good candidate for the `@property` decorator

**Comment by iknox-fa** [comment]
A few super minor details.. otherwise LGTM!

**Review by iknox-fa — APPROVED** [review]
👍

**Pull request merged**